src/Processors/QueryPlan/Optimizations/convertAnyJoinToSemiOrAntiJoin.cpp

```diff
@@ -0,0 +1,231 @@
+    }
+
+    for (const auto * input_node : dag.getInputs())
+    {
+        auto it = output_map.find(input_node->result_name);
+        if (it != output_map.end())
+            input[input_node] = it->second;
+    }
+
+    return input;
+}
+
+ColumnsWithTypeAndName partialyEvaluateActionsChain(const std::vector<ActionsDAG *> & actions_chain)
+{
+    ActionsDAG::IntermediateExecutionResult current_input;
+    for (size_t i = 0; i + 1 < actions_chain.size(); ++i)
+    {
+        auto partial_result = ActionsDAG::evaluatePartialResult(current_input, actions_chain[i]->getOutputs(), 1, false, true);
+
+        current_input = makeActionsInput(*actions_chain[i + 1], partial_result);
+    }
+
+    return ActionsDAG::evaluatePartialResult(current_input, actions_chain.back()->getOutputs(), 1, false, true);
+}
+
+enum class FilterResult
+{
+    UNKNOWN,
+    TRUE,
+    FALSE,
+};
+
+FilterResult getFilterResult(const ColumnWithTypeAndName & column)
+{
+    if (!column.column)
+        return FilterResult::UNKNOWN;
+
+    if (!column.type->canBeUsedInBooleanContext())
+        return FilterResult::UNKNOWN;
+
+    return column.column->getBool(0) ? FilterResult::TRUE : FilterResult::FALSE;
+}
+
+FilterResult filterResultForMatchedRows(const std::vector<ActionsDAG *> & pre_actions_chain, const ActionsDAG & filter_dag, const String & filter_column_name)
+{
+    auto pre_actions_output = partialyEvaluateActionsChain(pre_actions_chain);
+
+    ActionsDAG::IntermediateExecutionResult filter_input = makeActionsInput(filter_dag, pre_actions_output);
+    auto filter_output = ActionsDAG::evaluatePartialResult(filter_input, { filter_dag.tryFindInOutputs(filter_column_name) }, 1, false, true);
+
+    return getFilterResult(filter_output[0]);
+}
+
+FilterResult filterResultForNotMatchedRows(const ActionsDAG & filter_dag, const String & filter_column_name, const Block & input_stream_header)
+{
+    ActionsDAG::IntermediateExecutionResult filter_input;
+
+    /// Create constant columns with default values for inputs of the filter DAG
+    for (const auto * input : filter_dag.getInputs())
+    {
+        if (!input_stream_header.has(input->result_name))
+            continue;
+
+        if (input->column)
+        {
+            auto constant_column_with_type_and_name = ColumnWithTypeAndName{input->column, input->result_type, input->result_name};
+            filter_input.emplace(input, std::move(constant_column_with_type_and_name));
+            continue;
+        }
+
+        auto constant_column = input->result_type->createColumnConst(1, input->result_type->getDefault());
+        auto constant_column_with_type_and_name = ColumnWithTypeAndName{constant_column, input->result_type, input->result_name};
+        filter_input.emplace(input, std::move(constant_column_with_type_and_name));
+    }
+
+    auto filter_output = ActionsDAG::evaluatePartialResult(filter_input, { filter_dag.tryFindInOutputs(filter_column_name) }, 1, false, true);
```

Comment:
The sad part is that it can throw an exception on default values, e.g. `where intDiv(1, x)`.
And I don't know how to check it without throw/catch-ing

Reply:
AFAICS, it handles all the errors inside and doesn't throw any exceptions.